v2/datastream-to-spanner/terraform/samples/monitoring-alerting/modules/alerting/dataflow/dataflow.tf

```diff
@@ -0,0 +1,114 @@
+# Log-based metric for tracking Dataflow conversion errors separately
+resource "google_logging_metric" "dataflow_conversion_errors_metric" {
+  name = "dataflow_conversion_errors_metric"
+
+  # Filter to capture only conversion errors with severity of ERROR or higher for Dataflow jobs
+  filter = "resource.type=\"dataflow_job\" AND severity>=ERROR AND textPayload:\"conversion error\""
+
+  # Metric descriptor settings to define the metric type and value type
+  metric_descriptor {
+    metric_kind = "DELTA"  # Tracks change over time
+    value_type  = "INT64"  # Counts integer values
+  }
+
+}
+
+# Log-based metric for other (non-conversion) Dataflow errors
+resource "google_logging_metric" "dataflow_other_errors_metric" {
+  name        = "dataflow_other_errors_metric"
+  description = "Metric to track other Dataflow errors (excluding conversion errors)"
+
+  # Filter to capture all Dataflow errors except conversion errors
+  filter      = "resource.type=\"dataflow_job\" AND severity>=ERROR AND NOT textPayload:\"conversion error\""
+
+  metric_descriptor {
+    metric_kind = "DELTA"
+    value_type  = "INT64"
+  }
+
+}
+
+# Log-based metric for total Dataflow errors (both conversion and other errors)
+resource "google_logging_metric" "dataflow_total_errors_metric" {
+  name        = "dataflow_total_errors_metric"
+  description = "Metric to track the total number of Dataflow errors"
+
+  # Filter to capture all Dataflow errors with severity of ERROR or higher
+  filter      = "resource.type=\"dataflow_job\" AND severity>=ERROR"
+
+  metric_descriptor {
+    metric_kind = "DELTA"
+    value_type  = "INT64"
+  }
+
+}
```

Comment:
Why do we need log-based metrics? Aren't these metrics directly published to GCP in a counter?

An older (but similar) example of querying `Retryable Errors` from Monitoring - 

```
fetch dataflow_job
| metric 'dataflow.googleapis.com/job/user_counter'
| filter (resource.job_name == 'hb-dataflow-polltest-100tables-ef7e-3589')
| filter (metric.metric_name == 'Retryable errors')
| group_by 1m, [value_user_counter_mean: mean(value.user_counter)]
| every 1m
```